partiql-lang/src/test/kotlin/org/partiql/lang/planner/transforms/PartiQLSchemaInferencerTests.kt

```diff
@@ -714,9 +715,98 @@ class PartiQLSchemaInferencerTests {
+                expected = BagType(
+                    StructType(
+                        fields = listOf(
+                            StructType.Field("a", StaticType.INT),
+                            StructType.Field("a", StaticType.DECIMAL),
+                        ),
+                        contentClosed = true,
+                        constraints = setOf(TupleConstraint.Open(false), TupleConstraint.UniqueAttrs(true))
+                    )
+                )
+            ),
+            SuccessTestCase(
+                name = "LEFT JOIN ALL",
+                query = "SELECT * FROM <<{ 'a': 1 }>> AS t1 LEFT JOIN <<{ 'a': 2.0 }>> AS t2 ON t1.a = t2.a",
+                expected = BagType(
+                    StructType(
+                        fields = listOf(
+                            StructType.Field("a", StaticType.INT),
+                            StructType.Field("a", StaticType.DECIMAL),
+                        ),
+                        contentClosed = true,
+                        constraints = setOf(TupleConstraint.Open(false), TupleConstraint.UniqueAttrs(true))
+                    )
+                )
+            ),
+            SuccessTestCase(
+                name = "LEFT JOIN ALL",
+                query = """
+                    SELECT *
+                    FROM
+                            <<{ 'a': 1 }>> AS t1
+                        LEFT JOIN
+                            <<{ 'a': 2.0 }>> AS t2
+                        ON t1.a = t2.a
+                        LEFT JOIN
+                            <<{ 'a': 'hello, world' }>> AS t3
+                        ON t3.a = 'hello'
+                """,
+                expected = BagType(
+                    StructType(
+                        fields = listOf(
+                            StructType.Field("a", StaticType.INT),
+                            StructType.Field("a", StaticType.DECIMAL),
+                            StructType.Field("a", StaticType.STRING),
+                        ),
+                        contentClosed = true,
+                        constraints = setOf(TupleConstraint.Open(false), TupleConstraint.UniqueAttrs(true))
+                    )
+                )
+            ),
+            SuccessTestCase(
+                name = "Duplicate fields in struct",
+                query = """
+                    SELECT t.a AS a
+                    FROM <<
+                        { 'a': 1, 'a': 'hello' }
+                    >> AS t
+                """,
+                expected = BagType(
+                    StructType(
+                        fields = listOf(
+                            StructType.Field("a", unionOf(INT, STRING))
+                        ),
+                        contentClosed = true,
+                        constraints = setOf(TupleConstraint.Open(false), TupleConstraint.UniqueAttrs(true))
+                    )
+                )
+            ),
```

Comment:
Spec Page 4: 
> Unlike SQL, PartiQL allows the possibility of duplicate attribute names, in the interest of compatibility with nonstrict JSON/Ion datasets. However PartiQL does not encourage duplicate attribute names; navigation into tuples with the conventional dot notation (Section 4) is tuned to the assumption that the attribute names are unique.


Page 13:
> Tuple navigation in tuples with duplicate attributes When the tuple t has multiple attributes a, the tuple path navigation t.a will return the first instance of a. Note that for tuples whose order is defined by schema, this is well-defined, for unordered tuples, it is implementation defined which attribute is returned in permissive mode or an error in type checking mode, which is described in Section 4.1.

Reply:
If the intention here is to say that the plan will delegate the decision as which "a" to output(based on one of the comments above), then I personally think it is OK. But maybe worth to document this decision more formally.

Reply:
I think it makes sense to have the union of the types as part of logical plan. Basically there can be two possibilities:
1. Either we don't have a schema for the struct value which in that case there are no fields in the struct type.
2. We have a schema with duplicate keys and in that case, it's at least helpful to know what are the possible types for evaluation.

Considering this, @yliuuuu I can still see your concern with defining this more formally as a spec. (how we type our AST and represents the types in the plan) — maybe we can just open an issue in `partiql-spec` repo to track this?